### PR TITLE
feat: [PL-30092]: APIs to create and list of Audit Streaming Destinations

### PR DESCRIPTION
**Pull request by nishantsaini20**
## Describe your changes
This PR contains the following:
- openapi yaml and related files required for audit module as per spec first approach
- Implementation of two APIs (create and list) for streaming destinations
- UTs
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)

### CODE REVIEW

**Review by piyush-harness — APPROVED** [review]
> Reviewed 23 of 23 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: 23 file(s) to review. 1 of 2 approvals obtained.

File: platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        default: 0
+      description: 'Pagination page number strategy: Specify the page number within the paginated collection related to the number of items on each page.'
+    SearchTerm:
+      name: search_term
+      in: query
+      required: false
+      schema:
+        type: string
+      description: This would be used to filter resources having attributes matching the search term.
+    StreamingDestinationParam:
```
Comment (ashishsanodia-harness) [inline platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml:162]: Is this being used?

File: platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+      type: object
+      description: Streaming Destination response model
+      properties:
+        streamingDestination:
+          $ref: '#/components/schemas/StreamingDestinationDTO'
+        created:
+          type: integer
+          format: int64
+          description: Created time of the Streaming Destination
+        update:
```
Comment (mankrit-singh) [inline platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml:127]: nit: updated

File: platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        name:
+          type: string
+          description: Streaming Destination name
+        status:
+          type: string
+          description: Streaming Destination status
+          enum:
+            - ACTIVE
+            - INACTIVE
+        connectorRef:
```
Comment (mankrit-singh) [inline platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml:72]: Please rename it as connector_ref as part of standards

File: platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+      description: AwsS3 Streaming Destination model
+      x-discriminator-value: AWS_S3
+    StreamingDestinationResponse:
+      title: StreamingDestinationResponse
+      x-stoplight:
+        id: rulrehs9nk5a3
+      type: object
+      description: Streaming Destination response model
+      properties:
+        streamingDestination:
```
Comment (mankrit-singh) [inline platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml:121]: rename as streaming_destination

File: platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+                      name: example desitnation name
+                      status: ACTIVE
+                      connectorRef: account.example_connector
+                      spec:
+                        type: AWS_S3
+                        bucket: example_bucket
+                    created: 1670304106000
+                    update: 1670304106000
+      headers:
+        List:
```
Comment (mankrit-singh) [inline platform-service/modules/audit-service/contracts/openapi/v1/openapi.yaml:248]: rename as "Link" for consistency with other specs

**Review by mankrit-singh — CHANGES_REQUESTED** [review]

**Review by kapil-harness — CHANGES_REQUESTED** [review]
There are many class missing like  AwsS3StreamingDestinationSpecDTO,  AwsS3StreamingDestinationDTO,  StreamingDestinationResponse. Pls add these as w/o them its bit difficult to review.
Also, where is resource layer APIs and corresponding resource layer ITs.

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162#-:-NIi2u1iBEDejEBg72-N:b-86p4zy)* status: 24 file(s) to review. 2 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/entities/streaming/AwsS3StreamingDestination.java` line 31 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIhxFLPC1RLXqEr5Wmo:-NIhxFLPC1RLXqEr5Wmp:beh6sbr) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/entities/streaming/AwsS3StreamingDestination.java#L31)):*
> ```Java
> @FieldNameConstants(innerTypeName = "AwsS3StreamingDestinationKeys")
> @StoreIn(DbAliases.AUDITS)
> @Entity(value = "streamingDestinations", noClassnameStored = true)
> ```

this annotation is not required as it is a child to an entity class i.e. StreamingDestination
___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/entities/streaming/StreamingDestination.java` line 60 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIhwFhD0pnx8Vhlxunw:-NIhwFhD0pnx8Vhlxunx:b-f435fd) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/entities/streaming/StreamingDestination.java#L60)):*
> ```Java
> 
>   @Id @org.mongodb.morphia.annotations.Id String id;
>   @NotBlank @EntityIdentifier String identifier;
> ```

do we need 2 id's e.g. id and identifier ? Will one uuid won't suffice?
___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/AwsS3StreamingDestinationSpecDTOToEntityMapper.java` line 17 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIi0QTmCWxqaZ2u7Nds:-NIi0QTmCWxqaZ2u7Ndt:b8aqumz) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/AwsS3StreamingDestinationSpecDTOToEntityMapper.java#L17)):*
> ```Java
> 
>   public static AwsS3StreamingDestination toStreamingDestination(
>       AwsS3StreamingDestinationSpecDTO awsS3StreamingDestinationSpecDTO) {
> ```

not able to locate  AwsS3StreamingDestinationSpecDTO class in this PR.
___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/StreamingDestinationMapper.java` line 20 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIi16NQ-3YfXUm5-exp:-NIi16NRDb5UWEEVKZjZ:b-bbs6vx) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/StreamingDestinationMapper.java#L20)):*
> ```Java
>       String accountIdentifier, StreamingDestinationDTO streamingDestinationDTO) {
>     StreamingDestination streamingDestination = getStreamingDestination(streamingDestinationDTO.getSpec());
>     streamingDestination.setAccountIdentifier(accountIdentifier);
> ```

using a builder would be more graceful
___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/StreamingDestinationMapper.java` line 29 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIi-w0xByELgyg0Jgz0:-NIi-w0z7gUIKUuBAQFg:b-3aheze) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/StreamingDestinationMapper.java#L29)):*
> ```Java
>   }
> 
>   private StreamingDestination getStreamingDestination(StreamingDestinationSpecDTO streamingDestinationSpecDTO) {
> ```

neither able to find StreamingDestinationDTO nor StreamingDestinationSpecDTO class in this PR. is it not committed from your local ?
___
*[`platform-service/modules/audit-service/src/main/java/io/harness/serializer/morphia/NGAuditServiceMorphiaRegistrar.java` line 29 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIi2hKn2lcb-WHyLBR8:-NIi2hKn2lcb-WHyLBR9:b-gk5sin) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/serializer/morphia/NGAuditServiceMorphiaRegistrar.java#L29)):*
> ```Java
>     set.add(AuditSettings.class);
>     set.add(YamlDiffRecord.class);
>     set.add(StreamingDestination.class);
> ```

child entity class AWSStreamingDestination will also be added

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: 24 file(s) to review. 2 change requests. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/entities/streaming/AwsS3StreamingDestination.java` line 31 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIhxFLPC1RLXqEr5Wmo:-NIjyo8z5Nu7xaLWvdt9:bqjzsti) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/entities/streaming/AwsS3StreamingDestination.java#L31)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

this annotation is not required as it is a child to an entity class i.e. StreamingDestination
</blockquote></details>

Agree. Doesn't harm to have it though.
___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/entities/streaming/StreamingDestination.java` line 60 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIhwFhD0pnx8Vhlxunw:-NIjyvS300yPPFP2QPjv:bgiabfj) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/entities/streaming/StreamingDestination.java#L60)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

do we need 2 id's e.g. id and identifier ? Will one uuid won't suffice?
</blockquote></details>

Yes we need both,
id -> uuid
identifier -> user entered identifier
___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/AwsS3StreamingDestinationSpecDTOToEntityMapper.java` line 17 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIi0QTmCWxqaZ2u7Nds:-NIjzM5FAcjxqullGEJO:bs1lk8q) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/AwsS3StreamingDestinationSpecDTOToEntityMapper.java#L17)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

not able to locate  AwsS3StreamingDestinationSpecDTO class in this PR.
</blockquote></details>

This generated by bazel using openapi.yaml
___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/StreamingDestinationMapper.java` line 20 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIi16NQ-3YfXUm5-exp:-NIk-fvSCnof3VAil1il:b-8ril7o) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/StreamingDestinationMapper.java#L20)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

using a builder would be more graceful
</blockquote></details>

Builders with inheritance is comparatively more comple
___
*[`platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/StreamingDestinationMapper.java` line 29 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIi-w0xByELgyg0Jgz0:-NIk-wSI1B7krH6Aj3ix:b-3bc4hr) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/audit/mapper/streaming/StreamingDestinationMapper.java#L29)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

neither able to find StreamingDestinationDTO nor StreamingDestinationSpecDTO class in this PR. is it not committed from your local ?
</blockquote></details>

These are generated by bazel using openapi.yaml

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: 24 file(s) to review. 2 change requests. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`platform-service/modules/audit-service/src/main/java/io/harness/serializer/morphia/NGAuditServiceMorphiaRegistrar.java` line 29 at r3](https://reviewable.io/reviews/harness/harness-core/41162#-NIi2hKn2lcb-WHyLBR8:-NIk6XikAviF90BVdJAa:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/94404b8dc5a85ce06c06a8c84f107bf56b8dd141/platform-service/modules/audit-service/src/main/java/io/harness/serializer/morphia/NGAuditServiceMorphiaRegistrar.java#L29)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

child entity class AWSStreamingDestination will also be added
</blockquote></details>

Done.

**Review by mankrit-singh — COMMENTED** [review]
> Reviewed 14 of 23 files at r1, 6 of 8 files at r3, 1 of 2 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: 10 file(s) to review. 2 change requests. 7. unresolved discussions. 0 of 2 approvals obtained.

**Review by mankrit-singh — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: 10 file(s) to review. 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained.

**Review by kapil-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162#-:-NIlrLp0875Zxif5ivYD:bnfp4nl)* status: 10 file(s) to review. 1. unresolved discussions.

**Review by kapil-harness — DISMISSED** [review]
:lgtm:

> Reviewed 14 of 23 files at r1, 6 of 8 files at r3, 1 of 2 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162#-:-NIlt3MRBkpq04fQi9BZ:bnfp4nl)* status: 1. unresolved discussions.

**Review by gaurav-jain1 — DISMISSED** [review]
> Reviewed 14 of 23 files at r1, 6 of 8 files at r3, 1 of 2 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: 1. unresolved discussions.

**Review by ashishsanodia-harness — APPROVED** [review]
> Reviewed 14 of 23 files at r1, 6 of 8 files at r3, 1 of 2 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: 1. unresolved discussions.

**Review by ashishsanodia-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: :shipit: complete! ✓

**Review by mankrit-singh — APPROVED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Review by gaurav-jain1 — APPROVED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41162)* status: :shipit: complete! ✓

**Pull request merged**